### ISSUE TITLE
Clarify django version support

### ISSUE THREAD

**Issue by telenieko** [oscarmlage/django-cruds-adminlte#103]
Documentation says Django 1.8, setup.py says nothing, requirements.txt says 1.11.18

It looks that the bare minimum would be 2.0, otherwise utils.py cannot load (imports django.urls which is >= 2.0)

**Comment by telenieko** [comment]
Note that if 2.0 is the minimum, a note should be  put somewhere "this is a python 3 package only!" (tox.ini is currently misleading on that)

**Comment by oscarmlage** [comment]
IIRC there are two branches, the master one for _py2_ and _django1.x_ and another branch called `python3` where I was trying to clean up all the dependencies that were not properly ported to `python3` (plus  the _django_ version and some other external libraries). So yes, TBH right now there is a big mess tween `setup.py` and `requirements.txt`.

About the `tox.ini`, I've done a little test with _tox_ but in the end never finished the job, so don't trust so much in the ini.

The work in the `python3` branch was not complete at all, lot of stuff changing so your pull requests are very appreciated. I'll take a deep look to them, thank you so much.

**Comment by telenieko** [comment]
> About the `tox.ini`, I've done a little test with _tox_ but in the end never finished the job, so don't trust so much in the ini.
> 
> The work in the `python3` branch was not complete at all, lot of stuff changing so your pull requests are very appreciated. I'll take a deep look to them, thank you so much.

I created PR #104 which updates the tox.ini and cleans up the setup.py & requirements.txt a little. Also PR #105 adds CircleCI builds.

On versions, do you have any specific targets? Could we set Python >=3.7 and Django>=2.2 (lts) as desired minimums for the python3 branch? I'd update #104 accordingly

**Comment by oscarmlage** [comment]
> On versions, do you have any specific targets? Could we set Python >=3.7 and Django>=2.2 (lts) as desired minimums for the python3 branch? I'd update #104 accordingly

Yep, `Python >=3.7` and `Django>=2.2` seems to be a right spot to start on. Thank you @telenieko

**Comment by luisza** [comment]
Hi, I disaggree, Django 2.2 is not launched yet so last release is 2.1.9.  Also Django 2.1 supports Python 3.5, 3.6, and 3.7.   
This has a lot of sence because distributions on Linux (Debian and centos) comes with 3.5 o 3.6 so I Suggest start support to python 3.6 and 3.7 also if not represent a lot of work support 3.5.
We need to focus on LTS versions of Django so beta is not a good idea, but I also know that the effort of code to next LTS  has awards in the future

**Comment by telenieko** [comment]
> Hi, I disaggree, Django 2.2 is not launched yet

The latest official Django release is 2.2.2, 2.2 was released on April 1st and is an [LTS](https://www.djangoproject.com/download/) release.

Note that we are mainly talking about the python3 branch which is not released so there is a chance for a clean slate on backwards compatibility.

> This has a lot of sence because distributions on Linux (Debian and centos) comes with 3.5 o 3.6 so I Suggest start support to python 3.6 and 3.7 also if not represent a lot of work support 3.5.

Debian buster looks like it could be released sometime this summer (by looking at the current RC bug count) which will ship with 3.7.

I aknowledge though that Django 2.2 is Python >= 3.5, so if nothing prevents it we can do same.

Maybe we can have the test suite test against 35, 36 and 37 with Django 2.2 and leave the final decission on python versions for later down the reload (ie: before a relase). I've not yet seen any features used in the code that would prevent 3.5 from working.

> We need to focus on LTS versions of Django so beta is not a good idea, but I also know that the effort of code to next LTS has awards in the future

As stated above, 2.2 is the current LTS release (the previous LTS is 1.11) with the next LTS scheduled for 1Q2021.

**Comment by oscarmlage** [comment]
If I'm not wrong yesterday (_June 3, 2019_) they published `2.2.2` as the _latest official version_:

- https://docs.djangoproject.com/en/2.2/releases/2.2.2/

What do you think guys in having separate branches for `2.1` and `2.2`. Probably - in the end - lot of work can be used from one branch to the other. I'm open to other proposals but I agree that the project should move forward.

**Comment by telenieko** [comment]
> What do you think guys in having separate branches for `2.1` and `2.2`. Probably - in the end - lot of work can be used from one branch to the other. I'm open to other proposals but I agree that the project should move forward.

2.1 end of mainstream support was April 1st this year, end of extended support is December 2019, that is, in less than six months. That why I mentioned to jump directly to 2.2 (end of support April 2022)

**Comment by oscarmlage** [comment]
> 2.1 end of mainstream support was April 1st this year, end of extended support is December 2019, that is, in less than six months. That why I mentioned to jump directly to 2.2 (end of support April 2022)

Well, dunno how much time do you - guys - have to give some love to the project but I would be really happy if we could release a `Django-2.x` compatible version towards end of Summer, what leaves us Q4 with support for `2.1`. I would say to jump into `2.2` from the beginning _by default_ and, of course, feel free to consider maintain a separate branch for `2.1`.

Is that approach mostly ok with you too @luisza?

**Comment by telenieko** [comment]
> Well, dunno how much time do you - guys - have to give some love to the project but I would be really happy if we could release a `Django-2.x` compatible version towards end of Summer, what leaves us Q4 with support for `2.1`. I would say to jump into `2.2` from the beginning _by default_ and, of course, feel free to consider maintain a separate branch for `2.1`.

I'm using cruds-adminlte for an internal project for case management (we're a law firm). I myself don't know how much time I have, haha. Most of my patches are basicaly related to that as "I need them", or things I stumble upon like the three PRs I have open now :) which simply make life easier for everyone!

On #104 @oscarmlage:
> Also I would like to clear the way we should release according to the versions. I mean, stop working on improvements on Django-1.x compatible versions since 0.0.16 and move to Django-2 or work in separate branches or some other solution. What do you think? (probably we should discuss is as an issue, not in a PR)

I personaly focus on django 2 as the only project I'm using cruds-adminlte is django 2.2 + py3.7 but as said above, I don't foresee any python 3.7 specific features being currently used so python >=3.5 should not be hard.

**Comment by telenieko** [comment]
@oscarmlage the issue was closed with the merge, but maybe reopen it while we continue the debate on which versions to support (I don't have a reopen button!)

**Comment by luisza** [comment]
Hi, 
My fault, I was busy and checked faster the last version of Django. 
I we can focus in 2.2+ because are LTS, maybe when this is done we can check if work in 2.1 but as collateral effect.   With Python version we can support the officially supported in  Django 2.2 (3.5,3.6,3.7). 

>  if we could release a Django-2.x compatible version towards end of Summer

We can make a Milestone and put a tentative release date, so we can work to improve the 2.x compatibility.

**Comment by oscarmlage** [comment]
Hey guys, anyone is moving forward on this? I've just made a release right now. Maybe we can schedule some plans for the near future, what do you think?

**Comment by telenieko** [comment]
Hi,
Regarding the version minimums, I'd stick with what Django 2.2 already says as @luisza suggested. That is, Django >= 2.2, Python >= 3.5. Following django by default. If you agree I'll push minimum changes to setup.py

**Comment by oscarmlage** [comment]
It's perfectly ok with me, thanks @telenieko

**Comment by Guest007** [comment]
in setup.py we have django>=2 and django-ajax==2.3.7 
but django-ajax v2 is for django<2 and django-ajax v3 is for django>=2 
May be setup.py need to be fixed?

**Comment by telenieko** [comment]
> in setup.py we have django>=2 and django-ajax==2.3.7
> but django-ajax v2 is for django<2 and django-ajax v3 is for django>=2

Oh oh. Thanks for pointing this out. 

> May be setup.py need to be fixed?

If it just were that easy... But the AJAX stuff does not currently have test coverage thus changing the requirements without making sure the version upgrade works properly would not be very responsible... 

I have updated issue #15 accordingly, which is now a requirement to move forward on this ticket.
